cuML/src/solver/sgd.h

```diff
@@ -0,0 +1,267 @@
+/*
+ * Copyright (c) 2018, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include "ml_utils.h"
+#include <cuda_utils.h>
+#include <linalg/gemv.h>
+#include <stats/mean.h>
+#include <stats/mean_center.h>
+#include <linalg/add.h>
+#include <linalg/subtract.h>
+#include <linalg/norm.h>
+#include <linalg/eltwise.h>
+#include <linalg/unary_op.h>
+#include <linalg/cublas_wrappers.h>
+#include <matrix/math.h>
+#include <matrix/matrix.h>
+#include "glm/preprocess.h"
+#include "shuffle.h"
+#include <functions/linearReg.h>
+#include <functions/logisticReg.h>
+#include <functions/hinge.h>
+#include "learning_rate.h"
+
+namespace ML {
+namespace Solver {
+
+using namespace MLCommon;
+
+template<typename math_t>
+void sgdFit(math_t *input,
+		    int n_rows,
+		    int n_cols,
+		    math_t *labels,
+		    math_t *coef,
+		    math_t *intercept,
+		    bool fit_intercept,
+		    int batch_size,
+		    int epochs,
+		    ML::lr_type lr_type,
+		    math_t eta0,
+		    math_t power_t,
+		    ML::loss_funct loss,
+		    Functions::penalty penalty,
+		    math_t alpha,
+		    math_t l1_ratio,
+		    bool shuffle,
+		    math_t tol,
+		    int n_iter_no_change,
+		    cublasHandle_t cublas_handle,
+		    cusolverDnHandle_t cusolver_handle) {
+
+	ASSERT(n_cols > 1,
+			"Parameter n_cols: number of columns cannot be less than two");
+	ASSERT(n_rows > 1,
+			"Parameter n_rows: number of rows cannot be less than two");
+
+	math_t *mu_input = NULL;
+	math_t *mu_labels = NULL;
+	math_t *norm2_input = NULL;
+
+	if (fit_intercept) {
+		allocate(mu_input, n_cols);
+		allocate(mu_labels, 1);
+
+		GLM::preProcessData(input, n_rows, n_cols, labels, intercept, mu_input,
+				mu_labels, norm2_input, fit_intercept, false, cublas_handle,
+				cusolver_handle);
+	}
+
+	math_t *grads = NULL;
+	math_t *input_batch = NULL;
+	math_t *labels_batch = NULL;
+	math_t *loss_value = NULL;
+	int *indices = NULL;
+
+	allocate(grads, n_cols, true);
```

Comment:
I can't seem to find where the derivative wrt the bias would be stored.

Reply:
We process the data before learning the weights so we don't have to learn the bias.

Reply:
Is this how scikit does it?
This might only in special cases solve the join optimization \min_{w, b} loss(w, b).

Reply:
Check the SKL. Especially the OLS and Ridge. And look for what happens to intercept when the data is centered.

Reply:
@oyilmaz-nvidia Looking at ridge.py from sklearn here: https://github.com/scikit-learn/scikit-learn/blob/master/sklearn/linear_model/ridge.py#L435, they seem to be following the intercept-to-be-part-of-coefficients approach. And it seems efficient also to compute bias as part of computing coefficients itself.

Also looked at base.py's preprocess_data method here: https://github.com/scikit-learn/scikit-learn/blob/master/sklearn/linear_model/base.py#L92. I get that if intercept is not required then we should do mean-centering. But if user wants the intercept, then it is much easier (and faster) to not mean-center and compute this bias term as part of the coefficients itself, no?

Reply:
@teju85 Linear and Ridge regression use the _preprocess_data and _set_intercept functions in here https://github.com/scikit-learn/scikit-learn/blob/7389dba/sklearn/linear_model/base.py#L362 . If you look at the these two functions, we actually calculate the intercept. It's just a little bit different. I think this way is faster than calculating part of the coefficients approach because we mostly mean center the data (sometimes normalize as well). Rather than calculating the intercept at each iteration, it just does it once in the end. If you'd like to do it during the optimization, you can modify the functions files with an extra boolean parameter (like cal_intercept) to indicate calculating intercept or not at each iteration.